File: public/packages/dotnet/caching/distributed-redis/RedisCacheInvalidationBackplane.cs
// -----------------------------------------------------------------------
// <copyright file="RedisCacheInvalidationBackplane.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Caching.Distributed.Redis;

using System.Collections.Concurrent;
using DcsvIo.D2.Result;
using DcsvIo.D2.Utilities.Extensions;
using JetBrains.Annotations;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using StackExchange.Redis;

/// <summary>
/// Redis pub/sub-backed implementation of
/// <see cref="ICacheInvalidationBackplane"/>. Subscribes to a single
/// channel (<see cref="RedisCacheOptions.InvalidationChannel"/>) at
/// construction and dispatches incoming messages to every subscribed
/// handler. Handlers are invoked sequentially per message; exceptions
/// are isolated (one throwing handler doesn't break delivery to others).
/// </summary>
/// <remarks>
/// Universal "everyone acts" rule: no sender-ID filtering. Publishers
/// receive their own messages. Cost is bounded — see the abstraction's
/// XML doc and the lib README for the full semantics.
/// </remarks>
[MustDisposeResource(false)]
public sealed class RedisCacheInvalidationBackplane : ICacheInvalidationBackplane
{
    private readonly IConnectionMultiplexer r_redis;
    private readonly RedisCacheOptions r_options;
    private readonly ILogger<RedisCacheInvalidationBackplane> r_logger;
    private readonly ConcurrentDictionary<Guid, Subscription> r_subscriptions = new();
    private bool _disposed;

    /// <summary>Initializes a new <see cref="RedisCacheInvalidationBackplane"/>.</summary>
    /// <param name="redis">Redis connection multiplexer (singleton).</param>
    /// <param name="options">Cache options.</param>
    /// <param name="logger">Logger.</param>
    [MustDisposeResource(false)]
    public RedisCacheInvalidationBackplane(
        IConnectionMultiplexer redis,
        IOptions<RedisCacheOptions> options,
        ILogger<RedisCacheInvalidationBackplane> logger)
    {
        ArgumentNullException.ThrowIfNull(redis);
        ArgumentNullException.ThrowIfNull(options);
        ArgumentNullException.ThrowIfNull(logger);
        r_redis = redis;
        r_options = options.Value;
        r_logger = logger;

        // Subscribe to the channel once for the lifetime of this backplane
        // instance. Single Redis subscription dispatches to all handlers.
        var sub = r_redis.GetSubscriber();
        sub.Subscribe(
            RedisChannel.Literal(r_options.InvalidationChannel),
            (_, value) => OnMessageReceived(value!));
    }

    /// <inheritdoc />
    [MustDisposeResource]
    public IAsyncDisposable Subscribe(Func<string, CancellationToken, ValueTask> handler)
    {
        ArgumentNullException.ThrowIfNull(handler);
        ThrowIfDisposed();
        var id = Guid.NewGuid();
        var sub = new Subscription(id, handler, this);
        r_subscriptions[id] = sub;
        return sub;
    }

    /// <inheritdoc />
    public async ValueTask<D2Result> PublishInvalidationAsync(
        string key, CancellationToken ct = default)
    {
        if (key.Falsey()) return InputFailures.Required(nameof(key));
        try
        {
            await r_redis.GetSubscriber().PublishAsync(
                RedisChannel.Literal(r_options.InvalidationChannel),
                key);
            return D2Result.Ok();
        }
        catch (RedisException ex)
        {
            RedisCacheLog.RedisOpFailed(r_logger, "Backplane.Publish", ex.GetType().Name, key);
            return D2Result.ServiceUnavailable();
        }
    }

    /// <inheritdoc />
    public async ValueTask<D2Result> PublishInvalidationManyAsync(
        IReadOnlyCollection<string> keys, CancellationToken ct = default)
    {
        if (keys.Falsey()) return InputFailures.Required(nameof(keys));

        try
        {
            var sub = r_redis.GetSubscriber();
            var channel = RedisChannel.Literal(r_options.InvalidationChannel);
            var tasks = new List<Task<long>>(keys.Count);
            foreach (var key in keys)
                tasks.Add(sub.PublishAsync(channel, key));
            await Task.WhenAll(tasks);
            return D2Result.Ok();
        }
        catch (RedisException ex)
        {
            RedisCacheLog.RedisOpFailed(
                r_logger, "Backplane.PublishMany", ex.GetType().Name, $"{keys.Count} keys");
            return D2Result.ServiceUnavailable();
        }
    }

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        if (_disposed) return;
        _disposed = true;

        // Unsubscribe from the Redis channel so no more messages arrive.
        try
        {
            await r_redis.GetSubscriber().UnsubscribeAsync(
                RedisChannel.Literal(r_options.InvalidationChannel));
        }
        catch (RedisException)
        {
            // Best-effort on shutdown; swallow.
        }

        // Cancel any in-flight subscription work.
        foreach (var sub in r_subscriptions.Values)
            sub.SignalCancellation();
        r_subscriptions.Clear();
    }

    private void OnMessageReceived(string key)
    {
        // Dispatch to every active subscription. Each handler is run
        // sequentially with try/catch isolation. Async handlers are
        // awaited per-handler; this runs on the StackExchange.Redis
        // dispatch thread but Redis pub/sub doesn't backpressure us.
        foreach (var sub in r_subscriptions.Values)
        {
            // Fire-and-forget per subscription, with isolation. We don't
            // block the Redis dispatch thread on slow handlers.
            _ = InvokeHandlerAsync(sub, key);
        }
    }

    private async Task InvokeHandlerAsync(Subscription sub, string key)
    {
        try
        {
            await sub.Handler(key, sub.Token);
        }
        catch (OperationCanceledException) when (sub.Token.IsCancellationRequested)
        {
            // Subscription was disposed mid-handler; expected.
        }
        catch (Exception ex)
        {
            RedisCacheLog.BackplaneHandlerThrew(r_logger, ex.GetType().Name, key);
        }
    }

    private void Unsubscribe(Guid id) => r_subscriptions.TryRemove(id, out _);

    private void ThrowIfDisposed() => ObjectDisposedException.ThrowIf(_disposed, this);

    [MustDisposeResource(false)]
    private sealed class Subscription : IAsyncDisposable
    {
        private readonly RedisCacheInvalidationBackplane r_backplane;
        private readonly CancellationTokenSource r_cts = new();
        private bool _disposed;

        [MustDisposeResource(false)]
        internal Subscription(
            Guid id,
            Func<string, CancellationToken, ValueTask> handler,
            RedisCacheInvalidationBackplane backplane)
        {
            Id = id;
            Handler = handler;
            r_backplane = backplane;
        }

        internal Func<string, CancellationToken, ValueTask> Handler { get; }

        internal CancellationToken Token => r_cts.Token;

        private Guid Id { get; }

        public ValueTask DisposeAsync()
        {
            if (_disposed) return ValueTask.CompletedTask;
            _disposed = true;
            r_backplane.Unsubscribe(Id);
            r_cts.Cancel();
            r_cts.Dispose();
            return ValueTask.CompletedTask;
        }

        internal void SignalCancellation()
        {
            if (_disposed) return;
            r_cts.Cancel();
        }
    }
}

File: public/packages/typescript/caching/distributed-redis/src/redis-cache-invalidation-backplane.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import {
  InputFailures,
  type ICacheInvalidationBackplane,
} from "@dcsv-io/d2-caching-abstractions";
import type { ILogger } from "@dcsv-io/d2-logging";
import { sanitizedErrorRender } from "@dcsv-io/d2-logging";
import {
  canceled,
  ok,
  serviceUnavailable,
  type D2Result,
} from "@dcsv-io/d2-result";

/**
 * Closed-set backplane operation names for warn bindings (§21.11).
 */
export const BackplaneOp = {
  PUBLISH: "Backplane.Publish",
  PUBLISH_MANY: "Backplane.PublishMany",
} as const;

/** Closed-set type for {@link BackplaneOp} values. */
export type BackplaneOpName = (typeof BackplaneOp)[keyof typeof BackplaneOp];
import { falsey } from "@dcsv-io/d2-utilities";
import type Redis from "ioredis";
import { randomUUID } from "node:crypto";

import type { RedisCacheOptions } from "./redis-cache-options.js";

/**
 * Redis pub/sub implementation of {@link ICacheInvalidationBackplane}.
 *
 * **Dual-connection:** the host injects a **command** Redis client used
 * only for PUBLISH. At construction the backplane owns
 * `commandRedis.duplicate()` as the subscriber, starts channel subscribe
 * without an await in the sync constructor, and exposes
 * {@link ready} which resolves when channel subscription is established.
 *
 * Dispose unsubscribes and quits the **owned subscriber only** - never
 * the host command client. Publish after dispose does not throw (match
 * .NET). Handler-subscribe after dispose throws.
 *
 * Port `subscribe(handler)` remains sync and returns `AsyncDisposable`.
 */
export class RedisCacheInvalidationBackplane implements ICacheInvalidationBackplane {
  private readonly commandRedis: Redis;
  private readonly subscriber: Redis;
  private readonly channel: string;
  private readonly logger: ILogger;
  private readonly subscriptions = new Map<string, BackplaneSubscription>();
  private disposed = false;
  private channelReady = false;

  /**
   * Resolves when the owned-subscriber channel subscription is established.
   * Hosts and integration tests await this before delivery-dependent
   * publish/receive assertions. Construction stays synchronous.
   */
  readonly ready: Promise<void>;

  /**
   * @param commandRedis - Host-owned command client (publish path; never
   *   quit by this instance).
   * @param options - Cache options (channel name from `invalidationChannel`).
   * @param logger - Required structured logger.
   */
  constructor(
    commandRedis: Redis,
    options: RedisCacheOptions,
    logger: ILogger,
  ) {
    if (commandRedis == null) {
      throw new TypeError("commandRedis is required");
    }

    if (options == null) {
      throw new TypeError("options is required");
    }

    if (logger == null) {
      throw new TypeError("logger is required");
    }

    if (falsey(options.invalidationChannel)) {
      throw new RangeError(
        `RedisCacheOptions.invalidationChannel must be non-empty, ` +
          `got ${String(options.invalidationChannel)}.`,
      );
    }

    this.commandRedis = commandRedis;
    this.logger = logger;
    this.channel = options.invalidationChannel;
    this.subscriber = commandRedis.duplicate();

    this.subscriber.on("message", (ch, message) => {
      if (ch === this.channel) {
        this.onMessage(message);
      }
    });

    // Start channel subscribe without await in ctor - surface via ready.
    this.ready = this.subscriber.subscribe(this.channel).then(() => {
      this.channelReady = true;
    });
  }

  /**
   * Registers a handler. Sync (port contract) - not channel setup.
   *
   * @throws When the backplane is disposed or `handler` is nullish.
   */
  subscribe(
    handler: (key: string, signal?: AbortSignal) => void | Promise<void>,
  ): AsyncDisposable {
    if (handler == null) {
      throw new TypeError("handler is required");
    }

    this.throwIfDisposed();
    const id = randomUUID();
    const sub = new BackplaneSubscription(id, handler, this);
    this.subscriptions.set(id, sub);

    return sub;
  }

  /**
   * Publishes one invalidation key on the command client. Does not throw
   * when disposed (match .NET); may return `ok` or `serviceUnavailable`.
   */
  async publishInvalidation(
    key: string,
    signal?: AbortSignal,
  ): Promise<D2Result> {
    if (signal?.aborted === true) {
      return canceled();
    }

    if (falsey(key)) {
      return InputFailures.required("key");
    }

    try {
      await this.commandRedis.publish(this.channel, key);

      return ok();
    } catch (err) {
      this.logRedisOp(BackplaneOp.PUBLISH, err, key);

      return serviceUnavailable();
    }
  }

  /**
   * Bulk-publish on the command client. Does not throw when disposed.
   */
  async publishInvalidationMany(
    keys: readonly string[],
    signal?: AbortSignal,
  ): Promise<D2Result> {
    if (signal?.aborted === true) {
      return canceled();
    }

    if (falsey(keys)) {
      return InputFailures.required("keys");
    }

    try {
      await Promise.all(
        keys.map((key) => this.commandRedis.publish(this.channel, key)),
      );

      return ok();
    } catch (err) {
      this.logRedisOp(BackplaneOp.PUBLISH_MANY, err, `${keys.length} keys`);

      return serviceUnavailable();
    }
  }

  /**
   * Unsubscribes the channel on the owned subscriber, quits that
   * subscriber only, cancels handler subscriptions, and clears the map.
   * Idempotent. Never quits the host command client.
   */
  async dispose(): Promise<void> {
    if (this.disposed) {
      return;
    }

    this.disposed = true;

    try {
      await this.subscriber.unsubscribe(this.channel);
    } catch {
      // Best-effort on shutdown; swallow.
    }

    try {
      await this.subscriber.quit();
    } catch {
      // Best-effort on shutdown; swallow.
    }

    for (const sub of this.subscriptions.values()) {
      sub.signalCancellation();
    }

    this.subscriptions.clear();
  }

  /** Delegates to {@link dispose} for `await using` declarations. */
  async [Symbol.asyncDispose](): Promise<void> {
    await this.dispose();
  }

  /** @internal */
  unsubscribeHandler(id: string): void {
    this.subscriptions.delete(id);
  }

  private onMessage(key: string): void {
    if (!this.channelReady || this.disposed) {
      return;
    }

    for (const sub of this.subscriptions.values()) {
      void this.invokeHandler(sub, key);
    }
  }

  private async invokeHandler(
    sub: BackplaneSubscription,
    key: string,
  ): Promise<void> {
    try {
      await sub.handler(key, sub.signal);
    } catch (err) {
      if (sub.signal.aborted) {
        return;
      }

      const exceptionType = sanitizedErrorRender(err).name;
      this.logger.warn("Backplane handler threw", {
        exceptionType,
        key,
      });
    }
  }

  private throwIfDisposed(): void {
    if (this.disposed) {
      throw new Error("RedisCacheInvalidationBackplane is disposed.");
    }
  }

  private logRedisOp(
    op: BackplaneOpName,
    err: unknown,
    keyOrCount: string,
  ): void {
    const exceptionType = sanitizedErrorRender(err).name;
    this.logger.warn("Redis cache operation failed", {
      operation: op,
      exceptionType,
      keyOrCount,
    });
  }
}

class BackplaneSubscription implements AsyncDisposable {
  private readonly ac = new AbortController();
  private disposed = false;

  constructor(
    private readonly id: string,
    readonly handler: (
      key: string,
      signal?: AbortSignal,
    ) => void | Promise<void>,
    private readonly backplane: RedisCacheInvalidationBackplane,
  ) {}

  get signal(): AbortSignal {
    return this.ac.signal;
  }

  async [Symbol.asyncDispose](): Promise<void> {
    if (this.disposed) {
      return;
    }

    this.disposed = true;
    this.backplane.unsubscribeHandler(this.id);
    this.ac.abort();
  }

  /** Aborts the handler token (idempotent; safe after subscription dispose). */
  signalCancellation(): void {
    this.ac.abort();
  }
}
